src/Umbraco.Core/Scoping/Scope.cs

```diff
@@ -518,207 +512,235 @@ private static void TryFinally(int index, Action[] actions)
+                // The scope hasn't requested a lock yet, so we have to create a dict for it.
+                WriteLocks[instanceId] = new Dictionary<int, int>();
+                WriteLocks[instanceId][lockId] = 1;
             }
         }
 
         /// <summary>
-        /// Decrements the count of the WriteLocks with a specific lock object identifier we currently hold.
+        /// Increment the counter in ReadLocks for a specific scope instance and lock identifier.
         /// </summary>
-        /// <param name="lockId">Lock object identifier to decrement.</param>
-        /// <param name="amountToDecrement">Amount to decrement the lock count with</param>
-        public void DecrementWriteLock(int lockId, int amountToDecrement)
+        /// <remarks>
+        /// This will not lock on the lock object, ensure that you lock on _dictionaryLocker before calling this method.
+        /// </remarks>
+        /// <param name="lockId">Lock ID to increment.</param>
+        /// <param name="instanceId">Instance ID of the scope requesting the lock.</param>
+        private void IncrementReadLock(int lockId, Guid instanceId)
         {
-            // If we aren't the outermost scope, pass it on to the parent.
-            if (ParentScope != null)
+            // Since we've already checked that we're the parent in the WriteLockInner method, we don't need to check again.
+            // If it's the very first time a lock has been requested the ReadLocks dict hasn't been instantiated yet.
+            ReadLocks ??= new Dictionary<Guid, Dictionary<int, int>>();
+
+            var locksDictFound = ReadLocks.TryGetValue(instanceId, out var locksDict);
+            if (locksDictFound)
             {
-                ParentScope.DecrementWriteLock(lockId, amountToDecrement);
-                return;
+                locksDict.TryGetValue(lockId, out var value);
+                ReadLocks[instanceId][lockId] = value + 1;
             }
-
-            lock (_dictionaryLocker)
+            else
             {
-                WriteLocks[lockId] -= amountToDecrement;
+                // The scope hasn't requested a lock yet, so we have to create a dict for it.
+                ReadLocks[instanceId] = new Dictionary<int, int>();
+                ReadLocks[instanceId][lockId] = 1;
             }
         }
 
         /// <summary>
-        /// Increment the count of the read locks we've requested
+        /// Resets all read lock counters for a given scope instance to 0, signalling that the lock is no longer in use.
         /// </summary>
-        /// <remarks>
-        /// This should only be done on child scopes since it's then used to decrement the count later.
-        /// </remarks>
-        /// <param name="lockIds"></param>
-        private void IncrementRequestedReadLock(params int[] lockIds)
+        /// <param name="instanceId">Instance ID of the scope to clear of lock counters.</param>
+        private void ClearReadLocks(Guid instanceId)
```

Comment:
We can't pass the locks, because the initial invocation will come from one of the inner scopes where both the dicts will be null, so what ends up at the outmost scope will just be `null` for the dictionary.

Since we always need to clear both of the dicts regardless I combined the two methods into a single method that clears both dictionaries.